add Filename to MigrationPgError

davidmdm opened the pull request:
Simply put, when using `tern/migrate` to run migrations, if a migration fails the error is returned and the PgError message is printed to the user. It would be useful to show to the user which of their migration files caused the error.

Intended use:

```go
if err := m.Migrate(ctx); err != nil {
  var migrationErr migrate.MigrationPgError
  if errors.As(err, &migrationErr) {
    fmt.Fprintf(os.Stderr, "failed to perform migration(s): %s: %v\n", migrationErr.Filename, migrationErr.PgError)
    os.Exit(1)  
  }
  // handle error
}
```

davidmdm commented:
@jackc

jackc commented:
While migrations typically do come from files it isn't strictly required, so I would suggest that instead of `Filename` the field  should be named `Name` or `MigrationName`. 

As far as changing `MigrationPgError.Error()`, I think that is okay as long we can keep the existing error output for the CLI tool.

davidmdm commented:
@jackc 

Here is what I did in this PR:
- changed `Filename` -> `MigrationName`
- added the new `Error() string` method for MigrationPgError that included MigrationName in output
- added an `Unwrap() error` method for MigrationPgError is someone wants to get the PgError directly.

I don't believe that that this changes the output of the tern CLI.
If we look at the code, the error is never used directly.

```go
if err, ok := err.(migrate.MigrationPgError); ok {
			if err.Detail != "" {
				fmt.Fprintln(os.Stderr, "DETAIL:", err.Detail)
			}

			if err.Position != 0 {
				ele, err := ExtractErrorLine(err.Sql, int(err.Position))
				if err != nil {
					fmt.Fprintln(os.Stderr, err)
					os.Exit(1)
				}

				prefix := fmt.Sprintf("LINE %d: ", ele.LineNum)
				fmt.Fprintf(os.Stderr, "%s%s\n", prefix, ele.Text)

				padding := strings.Repeat(" ", len(prefix)+ele.ColumnNum-1)
				fmt.Fprintf(os.Stderr, "%s^\n", padding)
			}
		}
		os.Exit(1)
	}
```

jackc commented:
LGTM

davidmdm commented:
@jackc 

Can you tag master with this new commit? I want to use this new error context in my tern scripts.

Much appreciated :)

jackc commented:
Done.
